wagtail/contrib/modeladmin/tests/test_modeladmin_edit_handlers.py

```diff
@@ -0,0 +1,136 @@
+        # in Person.edit_handler
+        response = self.client.get('/admin/modeladmintest/person/create/')
+        self.assertIn('first_name', response.content.decode('UTF-8'))
+        self.assertIn('last_name', response.content.decode('UTF-8'))
+        self.assertIn('phone_number', response.content.decode('UTF-8'))
+        self.assertNotIn('address', response.content.decode('UTF-8'))
+        self.assertEqual(
+            [ii for ii in response.context['form'].fields],
+            ['first_name', 'last_name', 'phone_number']
+        )
+
+    def test_model_panels(self):
+        # tests that panel definition from model is used to create
+        # a model instance if no edit_handler is set on the model
+        response = self.client.post('/admin/modeladmintest/friend/create/', {
+            'first_name': "John",
+            'last_name': "Doe",
+            'address': "123 Main St Anytown",
+            'phone_number': "+123456789"
+        })
+        # Should redirect back to index
+        self.assertRedirects(response, '/admin/modeladmintest/friend/')
+
+        # Check that the friend was created
+        self.assertEqual(Friend.objects.filter(
+            first_name="John",
+            phone_number="+123456789"
+        ).count(), 1)
+
+        # verify that form fields are returned which have been defined
+        # in Friend.panels
+        response = self.client.get('/admin/modeladmintest/friend/create/')
+        self.assertIn('first_name', response.content.decode('UTF-8'))
+        self.assertNotIn('last_name', response.content.decode('UTF-8'))
+        self.assertIn('phone_number', response.content.decode('UTF-8'))
+        self.assertNotIn('address', response.content.decode('UTF-8'))
+        self.assertEqual(
+            [ii for ii in response.context['form'].fields],
+            ['first_name', 'phone_number']
+        )
+
+    def test_model_admin_edit_handler(self):
+        # tests that edit_handler definition from modeladmin is used to create
+        # a model instance if no edit_handler is set on the model
+        response = self.client.post('/admin/modeladmintest/visitor/create/', {
+            'first_name': "John",
+            'last_name': "Doe",
+            'address': "123 Main St Anytown",
+            'phone_number': "+123456789"
+        })
+        # Should redirect back to index
+        self.assertRedirects(response, '/admin/modeladmintest/visitor/')
+
+        # Check that the person was created
+        self.assertEqual(Visitor.objects.filter(
+            last_name="Doe",
+            phone_number="+123456789",
+            address="123 Main St Anytown"
+        ).count(), 1)
+
+        # verify that form fields are returned which have been defined
+        # in Person.edit_handler
+        response = self.client.get('/admin/modeladmintest/visitor/create/')
+
+        self.assertIn('last_name', response.content.decode('UTF-8'))
```

Comment:
I would also suggest that checking `response.context['form'].fields` exclusively in each of these tests would make your intentions quite a lot clearer... as the form is really the most indicative thing we have to test that the right `edit_handler` or `fields` were used.

Reply:
@ababic I agree and will fix it. This is also an approach I copied from `test_simple_modeladmin.TestCreateView.test_create`. But I agree that it makes the intention of the test case a lot clearer if we only check the form fields.

Reply:
@ababic fixed in https://github.com/wagtail/wagtail/pull/4900/commits/68f3b9829cede196926c48a694432eb7e2a0dbe8